src/middleware/mod.rs

```diff
@@ -2,38 +2,62 @@ use std::sync::Arc;
 
 use async_trait::async_trait;
 use futures::{future::BoxFuture, FutureExt};
+use jsonrpsee::core::{Error, JsonValue};
+use jsonrpsee::types::Params;
+use jsonrpsee::SubscriptionSink;
 
+pub mod cache;
 pub mod call;
 pub mod subscription;
 
+pub enum Request {
+    Call {
+        method: String,
+        params: Params<'static>,
+    },
+    Subscription {
+        subscribe: String,
+        params: Params<'static>,
+        unsubscribe: String,
+        sink: SubscriptionSink,
+    },
+}
+
 type NextFn<Request, Result> = Box<dyn FnOnce(Request) -> BoxFuture<'static, Result> + Send + Sync>;
 
 #[async_trait]
-pub trait Middleware<Request, Result: 'static>: Send + Sync {
-    async fn call(&self, request: Request, next: NextFn<Request, Result>) -> Result;
+pub trait Middleware: Send + Sync {
```

Comment:
I want to keep this generic so we can have different set of middleware for Calls and Subscriptions. We don't really want to implement middlewares that's able to handle both Calls and Subscriptions

Reply:
actually call and subscribe can be one middleware

Reply:
they can but I don't want them to be one. no middleware will handle both calls and subscriptions.

Reply:
makes more sense now?

Reply:
I still want this reverted. `Middleware<CallRequest>` and `Middleware<SubscriptionRequest>` are two different traits. The cache middleware should not care about subscriptions at all.